### PR TITLE
Added: InputKeyValueDropdown widget

### PR DESCRIPTION
**Pull request by vjosadika**
JIRA: https://cdap.atlassian.net/browse/PLUGIN-492

### CODE REVIEW

**Comment by CuriousVini** [comment]
@vjosadika Please add jira to the PR description

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  value: string | number;
+  label: string;
+}
+
+export type IDropdownOption = string | number | IComplexDropdown;
+
+interface IInputKeyValueDropdownProps extends IAbstractRowProps<typeof styles> {
+  keyPlaceholder?: string;
+  kvDelimiter?: string;
+  // dropdownOptions: IDropdownOption[];
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: nit: We can remove commented out code.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    if (typeof value === 'object') {
+      value = value.logicalType || value;
+    }
+    return value;
+  }
+
+  public renderInput = () => {
+    // const inputSchema = objectQuery(this.props.extraConfig, 'inputSchema');
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: Same here.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,181 @@
+import * as React from 'react';
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: This file needs license header.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: nit: 2020

File: cdap-ui/server/config/development/cdap.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
   "dashboard.ssl.disable.cert.check": "false",
   "ssl.external.enabled": "false",
   "router.server.address": "127.0.0.1",
-  "dashboard.bind.port": "11011",
+  "dashboard.bind.port": "11012",
```
Comment (ajainarayanan) [inline cdap-ui/server/config/development/cdap.json]: This is not needed.

File: cdap-ui/package.json [hunk 6]
```diff
@@ -125,7 +125,7 @@
     "main-bower-files": "2.13.3",
     "merge-stream": "2.0.0",
     "node-dir": "0.1.17",
-    "node-sass": "4.13.0",
+    "node-sass": "4.14.1",
```
Comment (ajainarayanan) [inline cdap-ui/package.json]: This change is not related to this widget right?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        const key = this.state.key;
+        const value = this.state.value;
+
+        const updatedValue = key.length > 0 ? [key, value].join(this.props.kvDelimiter) : '';
+        this.onChange(updatedValue);
+      }
+    );
+  };
+
+  public getFields(schemas: IStageSchema[], allowedTypes: string[]) {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: We seem to be calling `allowedTypes` always with `[]`. Is this really necessary?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+    if (typeof value === 'object') {
+      value = value.logicalType || value;
+    }
+    return value;
+  }
+
+  public renderInput = () => {
+    // const inputSchema = objectQuery(this.props.extraConfig, 'inputSchema');
+    const fieldValues = this.getFields(this.props.dropdownOptions, []);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: This doesn't look right. The `dropdownOptions` comes from `widgetProps` which means its a static one from the widget json (like placeholder or default value). 

Based on my understanding shouldn't this be list of fields from the input schema? If so then the commented out code seems right as this widget needs to show fields from current input schema. Did I understand that right?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      // tslint:disable-next-line: no-console
+      console.log('Error: Invalid JSON schema');
+    }
+    return fields;
+  }
+
+  // Function that checks if types contains a type that is in allowedTypes
+  // This is meant to handle nullable fields since a nullable string type is
+  // presented as ['string','null'].
+  public containsType(types: string[], allowedTypes: string[]) {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: Based on the usage of `allowedTypes` seems like both this and the `extractType` function is unnecessary?

**Review by ajainarayanan — COMMENTED** [review]
@vjosadika There seems to be some disconnect here. 

1. We already have a keyvalue dropdown widget. Apart from the ordering of components the new widget seems to be similar to the existing one. If this is what we need then we should have a widget that is a wrapper around existing keyvaluedropdown and add a property to existing keyvaluedropdown widget to configure the order of select and input element.
2. Based on the implementation of the new component, is the `dropdownOptions` coming from widget json? If so that seems wrong right? Everytime the plugin is configured it needs to get the dropdown options based on input schema. Correct me if my understanding here is not right.
3. If this widget is intended for use in pivotplugin for field aliases then wouldn't that require three inputs? 
  - values in the column
  - label for the column
  - alias of the column
If this is the case then I believe it is missing an input field?

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 10]
```diff
@@ -131,14 +143,34 @@ class KeyValueDropdownRow extends AbstractRow<IKeyValueDropdownRowProps, IKeyVal
...
-          {dropdownOptions.map((option) => {
-            return (
+          {data.map((option, index) =>
+            this.props.ordering !== OrderingEnum.VALUESFIRST ? (
               <MenuItem value={option.value} key={option.value} data-cy={`value-${option.value}`}>
                 {option.label}
               </MenuItem>
-            );
-          })}
+            ) : (
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: This shouldn't be changed. The option can either be string or it can be an object with `value` and `label`. The InputKeyValueDropdown component should pass in this options in the standard way.

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 11]
```diff
@@ -131,14 +143,34 @@ class KeyValueDropdownRow extends AbstractRow<IKeyValueDropdownRowProps, IKeyVal
           disabled={this.props.disabled}
           data-cy="value"
         >
-          {dropdownOptions.map((option) => {
-            return (
+          {data.map((option, index) =>
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: nit: IMO we shouldn't do this rename. The dropdownOptions shows that it is a list of options (array). It is hard to infer the type from the name `data`.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+interface IInputKeyValueDropdownProps extends IMultiRowProps<IInputKeyValueDropdownWidgetProps> {}
+
+interface IField {
+  name: string;
+  type: string[];
+}
+
+class InputKeyValueDropdownWidgetView extends AbstractMultiRowWidget<IInputKeyValueDropdownProps> {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: Can we rename this as `InputFieldDropdownKeyValue`? We already have `InputFieldDropdown` and `KeyValueDropdown`. This is basically a merge between them.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+          .map((field: IField) => field.name);
+      }
+    } catch {
+      // tslint:disable-next-line: no-console
+      console.log('Error: Invalid JSON schema');
+    }
+    return fields;
+  }
+
+  // Function that checks if types contains a type that is in allowedTypes
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: nit: Can you make this multiline comment?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    const dropdownOptions = objectQuery(this.props.extraConfig, 'inputSchema');
+    const fieldValues = this.getFields(dropdownOptions, []);
+
+    return (
+      <KeyValueDropdownRow
+        key={id}
+        value={this.values[id].value}
+        id={id}
+        index={index}
+        onChange={this.editRow}
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: Since we flip the ordering of the widget, the component ultimately returns `'key:value'` as string. To give an example,

- User configures the widget. Selects 'column1' from the input schema
- In the pivot textbox enters `v1, v2, v3`
- On change the component sets the value as `v1,v2,v3:column1`.

If we are flipping the ordering of the component we should handle the onchange as well if we expect the value to be in the same ordering, in this case `column1:v1, v2, v3`

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 15]
```diff
@@ -45,12 +45,19 @@ interface IComplexDropdown {
...
+}
+
 export type IDropdownOption = string | number | IComplexDropdown;
 
 interface IKeyValueDropdownRowProps extends IAbstractRowProps<typeof styles> {
   keyPlaceholder?: string;
   kvDelimiter?: string;
-  dropdownOptions: IDropdownOption[];
+  dropdownOptions?: IDropdownOption[];
+  inputSchemaFieldValues?: any;
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: This should be removed.

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 16]
```diff
@@ -45,12 +45,19 @@ interface IComplexDropdown {
...
+  ordering?: any;
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: This should be `OrderingEnum.KEYSFIRST | OrderingEnum.VALUESFIRST`

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 17]
```diff
@@ -64,6 +71,7 @@ class KeyValueDropdownRow extends AbstractRow<IKeyValueDropdownRowProps, IKeyVal
   public static defaultProps = {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx:71]: We need to have a default value for ordering which should be `OrderingEnum.KEYSFIRST`

**Review by ajainarayanan — COMMENTED** [review]
@vjosadika Had added some comments.

I still see changes related to `cdap.json` and the `node-sass` module changes. Could we remove them as they are not related to this PR?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx (hunk 7 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/InputKeyValueDropdown.tsx]: bump^

File: cdap-ui/server/config/development/cdap.json (hunk 5 above)
Comment (ajainarayanan) [reply cdap-ui/server/config/development/cdap.json]: Bump^

File: cdap-ui/app/cdap/components/AbstractWidget/AbstractWidgetFactory.tsx [hunk 18]
```diff
@@ -136,6 +137,7 @@ export const WIDGET_FACTORY = {
   dlp: DLPCustomWidget,
   'get-schema': GetSchemaWidget,
   'input-field-selector': InputFieldDropdown,
+  'input-keyvalue-dropdown': InputKeyValueDropdown,
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/AbstractWidgetFactory.tsx]: nit: This should be `input-field-keyvalue-dropdown`. We already have a `input-field-selector` indicating a selection of input schema fields. We should be consistent in the naming as this is what the plugin developers will use.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+} from 'components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow';
+
+import ThemeWrapper from 'components/ThemeWrapper';
+import { WIDGET_PROPTYPES } from 'components/AbstractWidget/constants';
+import { objectQuery } from 'services/helpers';
+
+interface IInputKeyValueDropdownWidgetProps extends IMultiRowWidgetProps {
+  'key-placeholder'?: string;
+  'kv-delimiter'?: string;
+  dropdownOptions: IDropdownOption[];
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: This is no longer being used right? The dropdown values are fields from input schema.

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+  delimiter?: string;
+  showDelimiter?: boolean;
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: This is not needed as well as we don't show the delimiter in the widget right?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+  );
+}
+
+(KeyValueDropdownWidget as any).propTypes = WIDGET_PROPTYPES;
+(KeyValueDropdownWidget as any).getWidgetAttributes = () => {
+  return {
+    'key-placeholder': { type: 'string', required: false },
+    'kv-delimiter': { type: 'string', required: false },
+    delimiter: { type: 'string', required: false },
+    dropdownOptions: { type: 'IDropdownOption[]', required: true },
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: Same here. Not needed.

File: cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx [hunk 22]
```diff
@@ -108,37 +116,59 @@ class KeyValueDropdownRow extends AbstractRow<IKeyValueDropdownRowProps, IKeyVal
...
           onKeyDown={this.handleKeyDown}
           disabled={this.props.disabled}
           inputRef={this.props.forwardedRef}
-          data-cy="key"
+          data-cy={props.type}
         />
+      );
+    };
 
+    const SelectField = (props) => {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: This won't work. When we define this as another component and pass in props we React will re-render this on each change as the props are different. That will make the focus lost from the current textbox. 

Was this change tested before being pushed?

Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: Please replace this change with just variables. We should not define this as another stateless react components. Also We shouldn't be mixing components (`props` and `this.props`)?

Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/AbstractWidget/KeyValueDropdownWidget/KeyValueDropdownRow.tsx]: Replace the snippet with
```
const inputType = this.props.ordering === OrderingEnum.KEYSFIRST ? 'key' : 'value';
    const InputField = (
      <Input
        classes={{ disabled: this.props.classes.disabled }}
        placeholder={this.props.keyPlaceholder}
        onChange={this.handleChange.bind(this, inputType)}
        value={this.props.ordering === OrderingEnum.KEYSFIRST ? this.state.key : this.state.value}
        autoFocus={this.props.autofocus}
        onKeyPress={this.handleKeyPress}
        onKeyDown={this.handleKeyDown}
        disabled={this.props.disabled}
        inputRef={this.props.forwardedRef}
        data-cy={inputType}
      />
    );

    const selectType = this.props.ordering === OrderingEnum.VALUESFIRST ? 'key' : 'value';
    const SelectField = (
      <Select
        classes={{ disabled: this.props.classes.disabled }}
        value={this.props.ordering === OrderingEnum.VALUESFIRST ? this.state.key : this.state.value}
        onChange={this.handleChange.bind(this, selectType)}
        displayEmpty={true}
        disabled={this.props.disabled}
        data-cy={selectType}
      >
        {dropdownOptions.map((option) => (
          <MenuItem
            value={option.value}
            key={option.value}
            data-cy={`${selectType}-${option.value}`}
          >
            {option.label}
          </MenuItem>
        ))}
      </Select>
    );```

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+      // tslint:disable-next-line: no-console
+      console.log('Error: Invalid JSON schema');
+    }
+    return fields;
+  }
+
+  /*   Function that checks if types contains a type that is in allowedTypes
+  This is meant to handle nullable fields since a nullable string type is
+  presented as ['string','null']. */
+  public containsType(types: string[], allowedTypes: string[]) {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: Can we remove these functions? These functions are not going to be called right (line 103)?

File: cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+          .map((field: IField) => field.name);
+      }
+    } catch {
+      // tslint:disable-next-line: no-console
+      console.log('Error: Invalid JSON schema');
+    }
+    return fields;
+  }
+
+  /*   Function that checks if types contains a type that is in allowedTypes
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/InputKeyValueDropdown/index.tsx]: If we need this function please format the multiline comment as,

```
/**
 *  Function that checks if types contains a type that is in allowedTypes
 * This is meant to handle nullable fields since a nullable string type is
 * presented as ['string','null'].
 */
```

**Comment by vjosadika** [comment]
@googlebot I signed it!

**Review by ajainarayanan — APPROVED** [review]
LGTM
